Add Alpha support for SME2 (#217)
ARM-software/acle: main/acle.md

    of a given ZA tile and a `_ver` suffix if they operate on vertical
    slices of a given ZA tile.

SME2 adds operations that work on groups of SVE vectors, ZA tile slices or
ZA array vectors. The intrinsics model this in the following way:

*   Multi-vector operands are groups of SVE data vectors, that use the same

>     tuple types as defined in the [SVE ACLE](#sve-vector-types), e.g.

Reviewer: Avoid Latin contractions.
e.g. -> for example.

    These immediates are required for indexing a double-vector group in the ZA
    array or to index a vector group of 2 tile-slices.

*   Every argument named `slice_x4_offset` must be an integer constant
    expression that is a multiple of 4.

    These immediates are required for indexing a quad-vector group in the ZA
    array or to index a vector group of 4 tile-slices.

*   Intrinsic functions that take a multi-vector operand may have additional
    suffixes to distinguish them from other forms for the same intrinsic:
    *   a `_single` suffix if they take one multi-vector operand and one
        (single) vector operand.
    *   a `_lane` suffix if they take one multi-vector operand and one
        indexed vector operand with an immediate to specify the indexed
        elements.

``` c
    __attributes__((arm_streaming, arm_shared_za))
    void svmla_za32[_s8]_vg4x2(uint32_t slice_base, uint64_t slice_x4_offset,
                               svint8x2_t zn, svint8x2_t zm);

    __attributes__((arm_streaming, arm_shared_za))
    void svmla[_single]_za32[_s8]_vg4x2(uint32_t slice_base,
                                        uint64_t slice_x4_offset, svint8x2_t zn,

>                                          svint8_t zm);

Reviewer 2: ```suggestion
                                        svint8_t zm);
```

    __attributes__((arm_streaming, arm_shared_za))
    void svmla_lane_za32[_s8]_vg4x2(uint32_t slice_base, uint64_t slice_x4_offset,
                                    svint8x2_t zn, svint8_t zm, uint64_t imm_idx);
```

### SME2 data-processing instructions.

#### ADD, SUB (store into ZA, single)

Multi-vector add/sub, storing into ZA

> The additional '_write' suffix indicates that the operation is not accumulating,

Reviewer 2: ```suggestion
The additional '_write' suffix indicates that the operation is not accumulating;
```

>  the result is written directly into ZA.

Reviewer 2: ```suggestion
the result is written directly into ZA.
```

  __attribute__((arm_streaming, arm_shared_za))
  void svadd_write[_single]_za32[_s32]_vg1x4(uint32_t slice_base,
                                             uint64_t slice_offset,
                                             svint32x4_t zn, svint32_t zm);
  
  
  // Variants are available for:
  //   _za32[_u32]
  //   _za64[_u64] (only if __ARM_FEATURE_SME_I16I64 != 0)
  __attribute__((arm_streaming, arm_shared_za))
  void svsub_write[_single]_za32[_u32]_vg1x2(uint32_t slice_base,
                                             uint64_t slice_offset,
                                             svuint32x2_t zn, svuint32_t zm);
  
  
  // Variants are available for:
  //   _za32[_u32]
  //   _za64[_u64] (only if __ARM_FEATURE_SME_I16I64 != 0)
  __attribute__((arm_streaming, arm_shared_za))
  void svsub_write[_single]_za32[_u32]_vg1x4(uint32_t slice_base,
                                             uint64_t slice_offset,
                                             svuint32x4_t zn, svuint32_t zm);
  
  ```

#### ADD, SUB (store into ZA, multi)

Multi-vector add/sub, storing into ZA

The additional '_write' suffix indicates that the operation is not accumulating,

>  the result is written directly into ZA.

Reviewer 2: ```suggestion
the result is written directly into ZA.
```

  svcount_t svwhilege_c8(int64_t rn, int64_t rm, uint64_t vl);
  
  
  // Variants are also available for _c16, _c32 and _c64
  __attribute__((arm_streaming))
  svcount_t svwhilegt_c8(int64_t rn, int64_t rm, uint64_t vl);
  
  
  // Variants are also available for _c16, _c32 and _c64
  __attribute__((arm_streaming))
  svcount_t svwhilehi_c8(uint64_t rn, uint64_t rm, uint64_t vl);
  
  
  // Variants are also available for _c16, _c32 and _c64
  __attribute__((arm_streaming))
  svcount_t svwhilehs_c8(uint64_t rn, uint64_t rm, uint64_t vl);
  
  
  // Variants are also available for _c16, _c32 and _c64
  __attribute__((arm_streaming))
  svcount_t svwhilele_c8(int64_t rn, int64_t rm, uint64_t vl);
  
  
  // Variants are also available for _c16, _c32 and _c64
  __attribute__((arm_streaming))
  svcount_t svwhilelo_c8(uint64_t rn, uint64_t rm, uint64_t vl);
  
  
  // Variants are also available for _c16, _c32 and _c64
  __attribute__((arm_streaming))
  svcount_t svwhilels_c8(uint64_t rn, uint64_t rm, uint64_t vl);
  
  
  // Variants are also available for _c16, _c32 and _c64
  __attribute__((arm_streaming))
  svcount_t svwhilelt_c8(int64_t rn, int64_t rm, uint64_t vl);
  
  ```

> #### WHILEGE, WHILEGT, WHILEHI, WHILEHS, WHILELE, WHILELO, WHILELS, WHILELT

Reviewer 2: Repeated heading.  Just removing it should be OK, so that there's two blocks of functions under the same heading.

  __attribute__((arm_streaming_compatible))
  svboolx2_t svwhilehs_b8_x2(uint64_t rn, uint64_t rm);
  
  
  // Variants are also available for _b16, _b32 and _b64
  __attribute__((arm_streaming_compatible))
  svboolx2_t svwhilele_b8_x2(int64_t rn, int64_t rm);
  
  
  // Variants are also available for _b16, _b32 and _b64
  __attribute__((arm_streaming_compatible))
  svboolx2_t svwhilelo_b8_x2(uint64_t rn, uint64_t rm);
  
  
  // Variants are also available for _b16, _b32 and _b64
  __attribute__((arm_streaming_compatible))
  svboolx2_t svwhilels_b8_x2(uint64_t rn, uint64_t rm);
  
  
  // Variants are also available for _b16, _b32 and _b64
  __attribute__((arm_streaming_compatible))
  svboolx2_t svwhilelt_b8_x2(int64_t rn, int64_t rm);
  
  ```

#### SUNPK, UUNPK

Multi-vector pack/unpack

``` c
  // Variants are also available for _u16_x2, _u32_x2, _s32_x2, _u64_x2 and _s64_x2
  __attribute__((arm_streaming))
  svint16x2_t svunpk[_s16_x2](svint8_t zn);
  
  
  // Variants are also available for _u16_x4, _u32_x4, _s32_x4, _u64_x4 and _s64_x4
  __attribute__((arm_streaming))
  svint16x4_t svunpk[_s16_x4](svint8x2_t zn);
  
  ```

#### ZIP, UZP

Multi-vector zip/unzip (2 vectors)

> The uzipq instructions operate on quad-words, but for convenience accept all element types.

Reviewer 2: ```suggestion
The `svuzpq` and `svzipq` intrinsics operate on quad-words, but for convenience accept
all element types.
```

``` c
  // Variants are also available for _u8, _u16, _s16, _f16, _bf16, _u32, _s32, _f32,
  // _u64, _s64 and _f64
  __attribute__((arm_streaming))
  svint8x2_t svuzp[_s8]_x2(svint8_t zn, svint8_t zm);
  
  
  // Variants are also available for _u8, _u16, _s16, _f16, _bf16, _u32, _s32, _f32,
  // _u64, _s64 and _f64
  __attribute__((arm_streaming))
  svint8x2_t svuzpq[_s8]_x2(svint8_t zn, svint8_t zm);
  
  
  // Variants are also available for _u8, _u16, _s16, _f16, _bf16, _u32, _s32, _f32,
  // _u64, _s64 and _f64
  __attribute__((arm_streaming))
  svint8x2_t svzip[_s8]_x2(svint8_t zn, svint8_t zm);
  
  
  // Variants are also available for _u8, _u16, _s16, _f16, _bf16, _u32, _s32, _f32,
  // _u64, _s64 and _f64
  __attribute__((arm_streaming))
  svint8x2_t svzipq[_s8]_x2(svint8_t zn, svint8_t zm);
  
  ```

> #### ZIP, UZP

Reviewer 2: Not sure it's worth spitting the `x2` and `x4` functions into separate blocks.  It's probably more convenient to the reader if a `x4` function comes immediately after any corresponding `x2` function, which seems to be the style followed in the other sections.

                                    svint32x2_t zn, svint32x2_t zm);
  
  
  // Variants are available for:
  //   _za32[_s32]
  //   _za32[_u32]
  //   _za64[_s64] (only if __ARM_FEATURE_SME_I16I64 != 0)
  //   _za64[_u64] (only if __ARM_FEATURE_SME_I16I64 != 0)
  __attribute__((arm_streaming, arm_shared_za))
  void svadd_write_za32[_s32]_vg1x4(uint32_t slice_base, uint64_t slice_offset,
                                    svint32x4_t zn, svint32x4_t zm);
  
  
  // Variants are available for:
  //   _za32[_u32]
  //   _za64[_u64] (only if __ARM_FEATURE_SME_I16I64 != 0)
  __attribute__((arm_streaming, arm_shared_za))
  void svsub_write_za32[_u32]_vg1x2(uint32_t slice_base, uint64_t slice_offset,
                                    svuint32x2_t zn, svuint32x2_t zm);
  
  
  // Variants are available for:
  //   _za32[_u32]
  //   _za64[_u64] (only if __ARM_FEATURE_SME_I16I64 != 0)
  __attribute__((arm_streaming, arm_shared_za))
  void svsub_write_za32[_u32]_vg1x4(uint32_t slice_base, uint64_t slice_offset,
                                    svuint32x4_t zn, svuint32x4_t zm);
  
  ```

#### ADD, SUB (vectors)

> Multi-vector add/sub

Reviewer 3: There is only assembly for ADD vectors.
Is that correct that the ACLE should also implement sub.
Even if there is no assembly instructions to map the intrinsic to?

Author: For the SVE ACLE we've shied away from doing this, because it hides a performance impact i.e. that the multi-vector ADD is one instruction, whereas SUB is multiple. I also don't believe there are any multi-vector negate instructions, which would make the performance difference more remarkable.

I should remove the `, SUB` from the section title here though, thanks for pointing that out!

Author: Resolved in https://github.com/ARM-software/acle/pull/217/commits/18c5da294a88c3a4fbb4c8d3301c5f71293f1aff

``` c
  // Variants are also available for _single_u8_x2, _single_s16_x2,
  // _single_u16_x2, _single_s32_x2, _single_u32_x2, _single_s64_x2 and
  // _single_u64_x2
  __attribute__((arm_streaming))
  svint8x2_t svadd[_single_s8_x2](svint8x2_t zdn, svint8_t zm);
  
  
  // Variants are also available for _single_u8_x4, _single_s16_x4,
  // _single_u16_x4, _single_s32_x4, _single_u32_x4, _single_s64_x4 and
  // _single_u64_x4
  __attribute__((arm_streaming))
  svint8x4_t svadd[_single_s8_x4](svint8x4_t zdn, svint8_t zm);
  
  ```

#### ADD, SUB, FADD, FSUB (accumulate into ZA)

Multi-vector add/sub and accumulate into ZA

``` c
  // Variants are available for:
  //   _za32[_f32]
  //   _za32[_s32]
  //   _za32[_u32]
  //   _za64[_f64] (only if __ARM_FEATURE_SME_F64F64 != 0)
  //   _za64[_s64] (only if __ARM_FEATURE_SME_I16I64 != 0)
  //   _za64[_u64] (only if __ARM_FEATURE_SME_I16I64 != 0)
  __attribute__((arm_streaming, arm_shared_za))

>   void svadd_za32[_f32]_vg1x2(uint32_t slice_base, uint64_t slice_offset,

Reviewer 4: Should we also add the f16 variants (FEAT_SME_F16F16) of fadd/fsub accumulate into ZA?

  svbfloat16_t svcvtn_bf16[_f32_x2](svfloat32x2_t zn);
  
  ```

#### FCVT, BFCVT, FCVTZS, FCVTZU, SCVTF, UCVTF

Multi-vector convert to/from floating-point.

``` c
  // Variants are also available for _f16[_f32_x2]
  __attribute__((arm_streaming))
  svbfloat16_t svcvt_bf16[_f32_x2](svfloat32x2_t zn);

  // Variants are also available for _f32[_u32_x2], _s32[_f32_x2] and _u32[_f32_x2]
  __attribute__((arm_streaming))
  svfloat32x2_t svcvt_f32[_s32_x2](svint32x2_t zn);
  
  
  // Variants are also available for _f32[_u32_x4], _s32[_f32_x4] and _u32[_f32_x4]
  __attribute__((arm_streaming))
  svfloat32x4_t svcvt_f32[_s32_x4](svint32x4_t zn);
  
  ```

#### SQCVT, SQCVTU, UQCVT

Multi-vector saturating extract narrow

``` c
  // Variants are also available for _u16[_s32_x2] and _u16[_u32_x2]
  __attribute__((arm_streaming))
  svint16_t svqcvt_s16[_s32_x2](svint32x2_t zn);
  
  
  // Variants are also available for _u8[_s32_x4], _u8[_u32_x4], _s16[_s64_x4],
  // _u16[_s64_x4] and _u16[_u64_x4]
  __attribute__((arm_streaming))
  svint8_t svqcvt_s8[_s32_x4](svint32x4_t zn);
  
  ```

#### SQCVTN, SQCVTUN, UQCVTN

Multi-vector saturating extract narrow and interleave

``` c

>   // Variants are also available for _u16[_s32_x2] and _u16[_u32_x2]

Reviewer 3: I don't see any instructions for qcvtn with x2 only x4
https://developer.arm.com/documentation/ddi0602/2022-09/SME-Instructions/UQCVTN--Multi-vector-unsigned-saturating-extract-narrow-and-interleave-?lang=en
https://developer.arm.com/documentation/ddi0602/2022-09/SME-Instructions/SQCVTUN--Multi-vector-signed-saturating-unsigned-extract-narrow-and-interleave-?lang=en
https://developer.arm.com/documentation/ddi0602/2022-09/SME-Instructions/SQCVTN--Multi-vector-signed-saturating-extract-narrow-and-interleave-?lang=en

Author: Hi @CarolineConcatto, the x2 forms are part of both SVE2p1 and SME2, and are therefore found under the 'SVE instructions' section of that page. They are defined here:

* https://developer.arm.com/documentation/ddi0602/2022-09/SVE-Instructions/SQCVTN--Signed-saturating-extract-narrow-and-interleave-?lang=en
* https://developer.arm.com/documentation/ddi0602/2022-09/SVE-Instructions/UQCVTN--Unsigned-saturating-extract-narrow-and-interleave-
* https://developer.arm.com/documentation/ddi0602/2022-09/SVE-Instructions/SQCVTUN--Signed-saturating-unsigned-extract-narrow-and-interleave-?lang=en

  __attribute__((arm_streaming, arm_shared_za))
  void svmlsl_lane_za32[_s8]_vg4x1(uint32_t slice_base, uint64_t slice_x4_offset,
                                   svint8_t zn, svint8_t zm, uint64_t imm_idx);
  
  
  // Variants are available for:
  //   _za32[_s8]
  //   _za32[_u8]
  //   _za64[_s16] (only if __ARM_FEATURE_SME_I16I64 != 0)
  //   _za64[_u16] (only if __ARM_FEATURE_SME_I16I64 != 0)
  __attribute__((arm_streaming, arm_shared_za))
  void svmlsl_lane_za32[_s8]_vg4x2(uint32_t slice_base, uint64_t slice_x4_offset,
                                   svint8x2_t zn, svint8_t zm, uint64_t imm_idx);
  
  
  // Variants are available for:
  //   _za32[_s8]
  //   _za32[_u8]
  //   _za64[_s16] (only if __ARM_FEATURE_SME_I16I64 != 0)
  //   _za64[_u16] (only if __ARM_FEATURE_SME_I16I64 != 0)
  __attribute__((arm_streaming, arm_shared_za))
  void svmlsl_lane_za32[_s8]_vg4x4(uint32_t slice_base, uint64_t slice_x4_offset,
                                   svint8x4_t zn, svint8_t zm, uint64_t imm_idx);
  
  ```

#### BFMLSLB, BFMLSLT

BFloat16 floating-point multiply-subtract long from single-precision (top/bottom)

``` c
  
  __attribute__((arm_streaming_compatible))

>   svfloat32_t svmlslb[_f32](svfloat32_t zda, svbfloat16_t zn, svbfloat16_t zm);

Reviewer 5: These definitions clash with some of the existing fp16->fp32 mlslb/t intrinsics added for SVE2.

Could these be renamed so that they are different from the fp16 intrinsics, possibly to `svbfmlslb`?

Author: Thanks for pointing out! I've done this in https://github.com/ARM-software/acle/pull/217/commits/3b8a6e7ad94764182c28863aadc6a036803af2bd

  svuint8x4_t svluti2_lane_zt[_u8]_x4(uint64_t zt, svuint8_t zn,
                                      uint64_t imm_idx);
  
  
  // Variants are also available for _zt[_u16] and _zt[_u32]
  __attribute__((arm_streaming, arm_shared_za, arm_preserves_za))
  svuint8_t svluti4_lane_zt[_u8](uint64_t zt, svuint8_t zn, uint64_t imm_idx);
  
  
  // Variants are also available for _zt[_u16] and _zt[_u32]
  __attribute__((arm_streaming, arm_shared_za, arm_preserves_za))
  svuint8x2_t svluti4_lane_zt[_u8]_x2(uint64_t zt, svuint8_t zn,
                                      uint64_t imm_idx);
  
  
  // Variants are also available for _zt[_u32]
  __attribute__((arm_streaming, arm_shared_za, arm_preserves_za))
  svuint16x4_t svluti4_lane_zt[_u16]_x4(uint64_t zt, svuint16_t zn,
                                        uint64_t imm_idx);
  
  ```

#### MOVA

Move multi-vectors to/from ZA

``` c
  // Variants are also available for _za8[_u8], _za16[_s16], _za16[_u16],
  // _za16[_f16], _za16[_bf16], _za32[_s32], _za32[_u32], _za32[_f32],
  // _za64[_s64], _za64[_u64] and _za64_[f64]
  __attribute__((arm_streaming, arm_shared_za, arm_preserves_za))

>   svint8x2_t svread_hor_za8[_s8]_vg2(uint32_t slice_base,

Reviewer 4: I think for elements 16-bits and above (za16-za64) we should be passing in a tile index for both the reads and writes. It seems to be only the 8-bit versions that are restricted to tile 0 only.

Author: Thanks, I've addressed that suggestion in https://github.com/ARM-software/acle/pull/217/commits/d7d716d65de5661ed730577d8c9b76a0bccb534c

                                uint64_t slice_x4_offset, svint8x4_t zn);
  
  
  // Variants are also available for _za64[_u64] and _za64[_f64]
  __attribute__((arm_streaming, arm_shared_za))
  void svwrite_za64[_s64]_vg1x2(uint32_t slice_base, uint64_t slice_offset,
                                svint64x2_t zn);
  
  
  // Variants are also available for _za64[_u64] and _za64[_f64]
  __attribute__((arm_streaming, arm_shared_za))
  void svwrite_za64[_s64]_vg1x4(uint32_t slice_base, uint64_t slice_offset,
                                svint64x4_t zn);
  
  ```

#### PTRUE

Initialise predicate-as-counter to all active

``` c
  // Variants are also available for _c16, _c32 and _c64
  __attribute__((arm_streaming))
  svcount_t svptrue_c8();
  
  ```

#### PEXT

Transform a predicate-as-counter to a predicate (pair).

``` c
  // Variants are also available for _c16, _c32 and _c64
  __attribute__((arm_streaming))
  svbool_t svpext_lane_c8(svcount_t pnn, uint64_t imm);
  
  
  // Variants are also available for _c16, _c32 and _c64
  __attribute__((arm_streaming))
  svboolx2_t svpext_lane_c8_x2(svcount_t pnn, uint64_t imm);
  
  ```

> #### PSEL

Reviewer 3: I believe this instructions is only available for SME:
https://developer.arm.com/documentation/ddi0602/2022-09/SVE-Instructions/PSEL--Predicate-select-between-predicate-register-or-all-false-?lang=en

Author: Well spotted, I'll remove them. I will keep the `svcount_t` variant of psel, because `svcount_t` was only added in SME2.

Author: Addressed in https://github.com/ARM-software/acle/pull/217/commits/18c5da294a88c3a4fbb4c8d3301c5f71293f1aff

Predicate select between predicate value or all-false

``` c
  // Variants are also available for _b16, _b32 and _b64
  __attribute__((arm_streaming_compatible))
  svbool_t svpsel_lane_b8(svbool_t pn, svbool_t pm, uint32_t idx, uint64_t imm);
  
  ```

#### CNTP

Set scalar to count from predicate-as-counter. ``vl`` is expected to be 2 or 4.

``` c
  // Variants are also available for _c16, _c32 and _c64
  __attribute__((arm_streaming))
  uint64_t svcntp_c8(svcount_t pnn, uint64_t vl);
  
  ```

> #### UCLAMP, SCLAMP, FCLAMP

Reviewer 3: SCLAMP and UCLAMP  single vector are SME
FCLAMP single vector is SME2
https://developer.arm.com/documentation/ddi0602/2022-09/SVE-Instructions/SCLAMP--Signed-clamp-to-minimum-maximum-vector-?lang=en
https://developer.arm.com/documentation/ddi0602/2022-09/SVE-Instructions/UCLAMP--Unsigned-clamp-to-minimum-maximum-vector-?lang=en
https://developer.arm.com/documentation/ddi0602/2022-09/SVE-Instructions/FCLAMP--Floating-point-clamp-to-minimum-maximum-number-?lang=en

Author: Sorry, I thought I already answered this question. 

The *multi-vector* UCLAMP/SCLAMP described in this section are SME2.

https://developer.arm.com/documentation/ddi0602/2022-09/SME-Instructions/UCLAMP--Multi-vector-unsigned-clamp-to-minimum-maximum-vector-?lang=en
https://developer.arm.com/documentation/ddi0602/2022-09/SME-Instructions/SCLAMP--Multi-vector-signed-clamp-to-minimum-maximum-vector-?lang=en

Multi-vector clamp to minimum/maximum vector

``` c
  // Variants are also available for _s8, _u8, _s16, _u16, _s32, _u32, _f32,
  // _s64, _u64 and _f64
  __attribute__((arm_streaming_compatible))
  svfloat16_t svclamp[_f16](svfloat16_t zd, svfloat16_t zn, svfloat16_t zm);
  
  
  // Variants are also available for _single_s8_x2, _single_u8_x2,
  // _single_s16_x2, _single_u16_x2, _single_s32_x2, _single_u32_x2,
  // _single_f32_x2, _single_s64_x2, _single_u64_x2 and _single_f64_x2
  __attribute__((arm_streaming))
  svfloat16x2_t svclamp[_single_f16_x2](svfloat16x2_t zd, svfloat16_t zn,
                                        svfloat16_t zm);
  
  
  // Variants are also available for _single_s8_x4, _single_u8_x4,
  // _single_s16_x4, _single_u16_x4, _single_s32_x4, _single_u32_x4,
  // _single_f32_x4, _single_s64_x4, _single_u64_x4 and _single_f64_x4
  __attribute__((arm_streaming))
  svfloat16x4_t svclamp[_single_f16_x4](svfloat16x4_t zd, svfloat16_t zn,
                                        svfloat16_t zm);
  
  ```

> #### REVD

Reviewer 3: I believe this instructions are only SME or SVE2.1. 
https://developer.arm.com/documentation/ddi0602/2022-09/SVE-Instructions/REVD--Reverse-64-bit-doublewords-in-elements--predicated--?lang=en

Author: Well spotted, I'll remove them.

Author: Addressed in https://github.com/ARM-software/acle/pull/217/commits/18c5da294a88c3a4fbb4c8d3301c5f71293f1aff

  
  
  // Variants are also available for _za8[_u8], _za16[_s16], _za16[_u16],
  // _za16[_f16], _za16[_bf16], _za32[_s32], _za32[_u32], _za32[_f32],
  // _za64[_s64], _za64[_u64] and _za64_[f64]
  __attribute__((arm_streaming, arm_shared_za))
  void svwrite_ver_za8[_s8]_vg4(uint64_t tile, uint32_t slice_base,
                                uint64_t slice_x4_offset, svint8x4_t zn);
  
  
  // Variants are also available for _za64[_u64] and _za64[_f64]
  __attribute__((arm_streaming, arm_shared_za))
  void svwrite_za64[_s64]_vg1x2(uint32_t slice_base, uint64_t slice_offset,
                                svint64x2_t zn);
  
  
  // Variants are also available for _za64[_u64] and _za64[_f64]
  __attribute__((arm_streaming, arm_shared_za))
  void svwrite_za64[_s64]_vg1x4(uint32_t slice_base, uint64_t slice_offset,
                                svint64x4_t zn);
  
  ```

#### PTRUE

Initialise predicate-as-counter to all active or all inactive.

``` c
  // Variants are also available for _c16, _c32 and _c64
  __attribute__((arm_streaming))
  svcount_t svptrue_c8();
  
  __attribute__((arm_streaming_compatible))

>   svcount_t svpfalse_c(void);

Reviewer 3: s/svpfalse_c/svpfalse_c8/

Author: I specifically chose `svpfalse_c` here to align with the `svpfalse_b` intrinsic. For svptrue, the `_b8/_b16_/_b32/_b64` suffix matters so the compiler can choose the right instruction to generate a value for the predicate mask (e.g. 11111111 (_b8), 10101010 (_b16), 10001000 (_b32), ..). For `pfalse` the value is always the same (all lanes inactive), hence the element type is not relevant and there is only `svpfalse_b()`.  The SME2 specification says:

> The canonical form of the all-false predicate using the predicate-as-counter encoding is bits [15:0] set to 0.

which is why `svcount_t svpfalse_c(void)` takes no element-type suffix either.

    suffixes to distinguish them from other forms for the same intrinsic:
    *   a `_single` suffix if they take one multi-vector operand and one
        (single) vector operand.
    *   a `_lane` suffix if they take one multi-vector operand and one
        indexed vector operand with an immediate to specify the indexed
        elements.

``` c
    __attributes__((arm_streaming, arm_shared_za))
    void svmlal_vg4x2_za32[_s8](uint32_t slice_base, uint64_t slice_x4_offset,
                                svint8x2_t zn, svint8x2_t zm);

    __attributes__((arm_streaming, arm_shared_za))
    void svmlal_vg4x2[_single]_za32[_s8](uint32_t slice_base,
                                         uint64_t slice_x4_offset, svint8x2_t zn,
                                         svint8_t zm);

    __attributes__((arm_streaming, arm_shared_za))
    void svmlal_vg4x2_lane_za32[_s8](uint32_t slice_base, uint64_t slice_x4_offset,
                                     svint8x2_t zn, svint8_t zm, uint64_t imm_idx);
```

### SME2 data-processing instructions.

#### ADD, SUB (store into ZA, single)

Multi-vector add/sub, storing into ZA

> The additional '_write' suffix indicates that the operation is not accumulating,

Reviewer 6: Did you mean to document the additional `_single` suffix this group has.

                                     svint8x4_t zn, svuint8_t zm,
                                     uint64_t imm_idx);
  
  
  
  __attribute__((arm_streaming, arm_shared_za))
  void svusvdot_lane_za32[_u8]_vg1x4(uint32_t slice_base, uint64_t slice_offset,
                                     svuint8x4_t zn, svint8_t zm,
                                     uint64_t imm_idx);
  
  
  // Variants are also available for _za32[_f16], _za32[_s16] and _za32[_u16]
  __attribute__((arm_streaming, arm_shared_za))
  void svvdot_lane_za32[_bf16]_vg1x2(uint32_t slice_base, uint64_t slice_offset,
                                     svbfloat16x2_t zn, svbfloat16_t zm,
                                     uint64_t imm_idx);
  
  
  // Variants are available for:
  //   _za32[_s8]
  //   _za32[_u8]
  //   _za64[_s16] (only if __ARM_FEATURE_SME_I16I64 != 0)
  //   _za64[_u16] (only if __ARM_FEATURE_SME_I16I64 != 0)
  __attribute__((arm_streaming, arm_shared_za))
  void svvdot_lane_za32[_s8]_vg1x4(uint32_t slice_base, uint64_t slice_offset,
                                   svint8x4_t zn, svint8_t zm, uint64_t imm_idx);
  
  ```

#### UMOPA, SMOPA, UMOPS, SMOPS

> Integer sum of outer products and accumulate/subtract (2-way and 4-way)

Reviewer 4: I think this should be 

Integer sum of outer products and accumulate/subtract (2-way)

because the 4-way variants were added as part of SME1.

``` c
  // Variants are also available for _za32[_u16]
  __attribute__((arm_streaming, arm_shared_za))

>   void svmopa_za32[_s16]_m(uint64_t tile, svbool_t pn, svbool_t pm, svint16_t zn,

Reviewer 4: I think the two builtins listed here refer only to the 2-way variants, but we also have the 4-way variants:

smvopa_za32_s8_m
smvopa_za64_s16_m (FEAT_SME_I16I64)
etc.

Reviewer 4: Please ignore the above comment! I was getting confused - the 4-way variants were added as part of SME1.

  
  
  // Variants are also available for _f32 and _f64
  __attribute__((arm_streaming))
  svfloat16x4_t svminnm[_f16][_x4](svfloat16x4_t zdn, svfloat16x4_t zm);
  
  ```

#### FRINTA, FRINTM, FRINTN, FRINTP

Multi-vector floating-point round to integral value

``` c
  
  __attribute__((arm_streaming))
  svfloat32x2_t svrinta[_f32][_x2](svfloat32x2_t zn);
  
  
  
  __attribute__((arm_streaming))
  svfloat32x4_t svrinta[_f32][_x4](svfloat32x4_t zn);
  
  
  
  __attribute__((arm_streaming))
  svfloat32x2_t svrintm[_f32][_x2](svfloat32x2_t zn);
  
  
  
  __attribute__((arm_streaming))
  svfloat32x4_t svrintm[_f32][_x4](svfloat32x4_t zn);
  
  
  
  __attribute__((arm_streaming))
  svfloat32x2_t svrintn[_f32][_x2](svfloat32x2_t zn);
  
  
  
  __attribute__((arm_streaming))
  svfloat32x4_t svrintn[_f32][_x4](svfloat32x4_t zn);
  
  
  
  __attribute__((arm_streaming))
  svfloat32x2_t svrintp[_f32][_x2](svfloat32x2_t zn);
  
  
  
  __attribute__((arm_streaming))
  svfloat32x4_t svrintp[_f32][_x4](svfloat32x4_t zn);
  
  ```

#### LD1B, LD1D, LD1H, LD1W

Contiguous load to multi-vector

``` c
  
  __attribute__((arm_streaming))

>   svuint8x2_t svld1b_vnum[_u8]_x2(svcount_t png, const uint8_t *rn,

Reviewer 2: Also, the convention for SVE has been to drop the access size from the instruction name unless the load is performing an extension (in which case the suffix is `sb` for signed extension from bytes and `ub` for unsigned extension from bytes).

                                svint64x2_t zn);
  
  
  // Variants are also available for _za64[_u64]
  __attribute__((arm_streaming, arm_shared_za))
  void svwrite_za64[_s64]_vg1x4(uint32_t slice_base, uint64_t slice_offset,
                                svint64x4_t zn);
  
  ```

#### PTRUE

Initialise predicate-as-counter to all active

``` c
  // Variants are also available for _c16, _c32 and _c64
  __attribute__((arm_streaming))
  svcount_t svptrue_c8();
  
  ```

#### PEXT

Transform a predicate-as-counter to a predicate (pair).

``` c
  // Variants are also available for _c16, _c32 and _c64
  __attribute__((arm_streaming))
  svbool_t svpext_lane_c8(svcount_t pnn, uint64_t imm);
  
  
  // Variants are also available for _c16, _c32 and _c64
  __attribute__((arm_streaming))
  svboolx2_t svpext_lane_c8_x2(svcount_t pnn, uint64_t imm);
  
  ```

#### PSEL

Predicate select between predicate value or all-false

``` c
  // Variants are also available for _b16, _b32 and _b64
  __attribute__((arm_streaming_compatible))
  svbool_t svpsel_lane_b8(svbool_t pn, svbool_t pm, uint32_t idx, uint64_t imm);
  
  ```

#### CNTP

Set scalar to count from predicate-as-counter

``` c

>   enum sv_vgkind {

Author: That's a good suggestion, I've made the change.

  
  
  // Variants are also available for _u16, _u32 and _u64
  __attribute__((arm_streaming))
  svuint8x4_t svsel[_u8][_x4](svcount_t png, svuint8x4_t zn, svuint8x4_t zm);
  
  ```

#### URSHL, SRSHL (single)

Multi-vector rounding shift left

``` c
  // Variants are also available for _s16, _s32, _s64, _u8, _u16, _u32 and _u64
  __attribute__((arm_streaming))
  svint8x2_t svrshl[_single][_s8][_x2](svint8x2_t zdn, svint8_t zm);
  
  
  // Variants are also available for _s16, _s32, _s64, _u8, _u16, _u32 and _u64
  __attribute__((arm_streaming))
  svint8x4_t svrshl[_single][_s8][_x4](svint8x4_t zdn, svint8_t zm);
  
  ```

#### URSHL, SRSHL (multi)

Multi-vector rounding shift left

``` c
  // Variants are also available for _s16, _s32, _s64, _u8, _u16, _u32 and _u64
  __attribute__((arm_streaming))
  svint8x2_t svrshl[_s8][_x2](svint8x2_t zdn, svint8x2_t zm);
  
  
  // Variants are also available for _s16, _s32, _s64, _u8, _u16, _u32 and _u64
  __attribute__((arm_streaming))
  svint8x4_t svrshl[_s8][_x4](svint8x4_t zdn, svint8x4_t zm);
  
  ```

#### SQRSHR, UQRSHR

Multi-vector saturating rounding shift right narrow

``` c
  
  __attribute__((arm_streaming))

>   svint16_t svqrshr[_s16][_x2](svint32x2_t zn, uint64_t imm_1_8);

Reviewer 5: The immediate operand for the "multi-vector saturating rounding shift right" intrinsics are all named _imm_1_8_, though I believe the immediates are all in the range of 1 to number of bits per source element for these instructions. Should the name be changed to reflect this?

Author: Fair point. I've generalised the name to just `imm` to avoid the issue entirely.

  
  
  // Variants are also available for _b16, _b32 and _b64
  __attribute__((arm_streaming_compatible))
  svboolx2_t svwhilele_b8_x2(int64_t rn, int64_t rm);
  
  
  // Variants are also available for _b16, _b32 and _b64
  __attribute__((arm_streaming_compatible))
  svboolx2_t svwhilelo_b8_x2(uint64_t rn, uint64_t rm);
  
  
  // Variants are also available for _b16, _b32 and _b64
  __attribute__((arm_streaming_compatible))
  svboolx2_t svwhilels_b8_x2(uint64_t rn, uint64_t rm);
  
  
  // Variants are also available for _b16, _b32 and _b64
  __attribute__((arm_streaming_compatible))
  svboolx2_t svwhilelt_b8_x2(int64_t rn, int64_t rm);
  
  ```

#### SUNPK, UUNPK

Multi-vector pack/unpack

``` c
  // Variants are also available for _s32, _s64, _u16, _u32 and _u64
  __attribute__((arm_streaming))
  svint16x2_t svunpk[_s16][_x2](svint8_t zn);
  
  
  // Variants are also available for _s32, _s64, _u16, _u32 and _u64
  __attribute__((arm_streaming))
  svint16x4_t svunpk[_s16][_x4](svint8x2_t zn);
  
  ```

#### ZIP, UZP

Multi-vector zip/unzip (2 vectors)

The uzipq instructions operate on quad-words, but for convenience accept all element types.

``` c

>   // Variants are also available for _s16, _s32, _s64, _u8, _u16, _u32 and _u64

Reviewer 2: The list should include float types too.

  
  
  // Variants are also available for _single_u8_x4, _single_s16_x4,
  // _single_u16_x4, _single_s32_x4, _single_u32_x4, _single_s64_x4 and
  // _single_u64_x4
  __attribute__((arm_streaming))
  svint8x4_t svadd[_single_s8_x4](svint8x4_t zdn, svint8_t zm);
  
  ```

#### ADD, SUB, FADD, FSUB (accumulate into ZA)

Multi-vector add/sub and accumulate into ZA

``` c
  // Variants are available for:
  //   _za32[_f32]
  //   _za32[_s32]
  //   _za32[_u32]
  //   _za64[_f64] (only if __ARM_FEATURE_SME_F64F64 != 0)
  //   _za64[_s64] (only if __ARM_FEATURE_SME_I16I64 != 0)
  //   _za64[_u64] (only if __ARM_FEATURE_SME_I16I64 != 0)
  __attribute__((arm_streaming, arm_shared_za))
  void svadd_za32[_f32]_vg1x2(uint32_t slice, svfloat32x2_t zm);
  
  
  // Variants are available for:
  //   _za32[_f32]
  //   _za32[_s32]
  //   _za32[_u32]
  //   _za64[_f64] (only if __ARM_FEATURE_SME_F64F64 != 0)
  //   _za64[_s64] (only if __ARM_FEATURE_SME_I16I64 != 0)
  //   _za64[_u64] (only if __ARM_FEATURE_SME_I16I64 != 0)
  __attribute__((arm_streaming, arm_shared_za))
  void svadd_za32[_f32]_vg1x4(uint32_t slice, svfloat32x4_t zm);
  
  
  // Variants are available for:

>   //   _za32[_f32]

Reviewer 2: I think these svsubs should have s32 variants too, since the underlying instruction is sign-agnostic.

  __attribute__((arm_streaming))
  svfloat32x4_t svcvt_f32[_s32_x4](svint32x4_t zn);
  
  ```

#### SQCVT, SQCVTU, UQCVT

Multi-vector saturating extract narrow

``` c
  // Variants are also available for _u16[_s32_x2] and _u16[_u32_x2]
  __attribute__((arm_streaming))
  svint16_t svqcvt_s16[_s32_x2](svint32x2_t zn);
  
  
  // Variants are also available for _u8[_s32_x4], _u8[_u32_x4], _s16[_s64_x4],
  // _u16[_s64_x4] and _u16[_u64_x4]
  __attribute__((arm_streaming))
  svint8_t svqcvt_s8[_s32_x4](svint32x4_t zn);
  
  ```

#### SQCVTN, SQCVTUN, UQCVTN

Multi-vector saturating extract narrow and interleave

``` c
  // Variants are also available for _u16[_s32_x2] and _u16[_u32_x2]
  __attribute__((arm_streaming_compatible))
  svint16_t svqcvtn_s16[_s32_x2](svint32x2_t zn);
  
  
  // Variants are also available for _u8[_s32_x4], _u8[_u32_x4], _s16[_s64_x4],
  // _u16[_s64_x4] and _u16[_u64_x4]
  __attribute__((arm_streaming))
  svint8_t svqcvtn_s8[_s32_x4](svint32x4_t zn);
  
  ```

#### UDOT, SDOT, FDOT (vectors)

Multi-vector dot-product (2-way and 4-way)

``` c
  // Variants are also available for _s32_s16_s16 and _u32_u16_u16
  __attribute__((arm_streaming_compatible))

>   svfloat32_t svdot[_f32_f16_f16](svfloat32_t zda, svfloat16_t zn, svfloat16_t zm);

Reviewer 2: I wonder whether we should support `_n` versions of these functions too.  The corresponding 4-way functions allow things like:
```
svint32_t f(svint32_t x, svint8_t y) {
    return svdot(x, y, 1)
}
```
so it would be more consistent to also allow:
```
svint32_t f(svint32_t x, svint16_t y) {
    return svdot(x, y, 1)
}
```
The same thing applies to other arithmetic functions that have a single vector as their final operand.  E.g. perhaps the `[_single]` functions should also have an `[_n]` form that takes a scalar rather than a vector.  (The `[_n]` suffix would then replace the `[_single]` suffix, rather than add to it.)

Reviewer 2: On second thoughts, the current `_n` versions of the `svdot` intrinsics rely on the fact that the third argument's type is predictable from the first two arguments' types.  In contrast, having three type suffixes allows the third argument's type to vary arbitrarily from the other two's, although in practice the most useful difference is probably going to be signed vs. unsigned.  So having `_n` forms of `svdot[_t0_t1_t2]` would mean using a scalar value to infer a type suffix for `t2`.  We've deliberately avoided inferring type suffixes from scalars in other contexts (e.g. `svdup`) because of the usual integer promotions.

So I suppose it's better to leave the `_n` form out after all.

  
  ```

#### UDOT, SDOT, FDOT (indexed)

Multi-vector dot-product (2-way and 4-way)

``` c
  // Variants are also available for _s32 and _u32
  __attribute__((arm_streaming_compatible))

>   svfloat32_t svdot_lane[_f32](svfloat32_t zda, svfloat16_t zn, svfloat16_t zm,

Reviewer 2: There are already 4-way functions with the names `svdot_lane[_s32]` and `svdot_lane[_u32]`, so the new ones will need to have an extra suffix, like for the vector-vector version above.  In future, we'll also need a 4-way version of FDOT for:

* https://developer.arm.com/documentation/ddi0602/2023-09/SIMD-FP-Instructions/FDOT--4-way--by-element---8-bit-floating-point-dot-product-to-single-precision--vector--by-element--?lang=en

So I think we'll need an extra suffix for the `f32` case too.

                               uint64_t imm_idx);
  
  ```

#### FDOT, BFDOT, SUDOT, USDOT, SDOT, UDOT (store into ZA, single)

Multi-vector dot-product (2-way and 4-way)

``` c
  // Variants are available for:
  //   _za32[_bf16]
  //   _za32[_f16]
  //   _za32[_s8]
  //   _za32[_s16]
  //   _za32[_u8]
  //   _za32[_u16]
  //   _za64[_s16] (only if __ARM_FEATURE_SME_I16I64 != 0)
  //   _za64[_u16] (only if __ARM_FEATURE_SME_I16I64 != 0)
  __attribute__((arm_streaming, arm_shared_za))
  void svdot[_single]_za32[_bf16]_vg1x2(uint32_t slice, svbfloat16x2_t zn,
                                        svbfloat16_t zm);
  
  
  // Variants are available for:
  //   _za32[_bf16]
  //   _za32[_f16]
  //   _za32[_s8]
  //   _za32[_s16]
  //   _za32[_u8]
  //   _za32[_u16]
  //   _za64[_s16] (only if __ARM_FEATURE_SME_I16I64 != 0)
  //   _za64[_u16] (only if __ARM_FEATURE_SME_I16I64 != 0)
  __attribute__((arm_streaming, arm_shared_za))
  void svdot[_single]_za32[_bf16]_vg1x4(uint32_t slice,
                                        svbfloat16x4_t zn, svbfloat16_t zm);
  
  
  __attribute__((arm_streaming, arm_shared_za))

>   void svsudot[_single]_za32[_s8]_vg1x2(uint32_t slice, svint8x2_t zn,

Reviewer 2: I suppose this raises a policy decision: do we want to continue to encode sign differences in the mnemonic, as for Advanced SIMD and SVE?  Or do we want to use suffixes?  If we want to use the mnemonic, there's probably no need for three type suffixes on line 9925.  Just two should be enough (which would make the `_n` forms feasible again).  If instead we switch to using suffixes, then we should probably do the same here too (`svdot[_single]_za32[_s8_u8]_vg1x2`).

  
  
  // Variants are available for:
  //   _za32[_f32]
  //   _za64[_f64] (only if __ARM_FEATURE_SME_F64F64 != 0)
  __attribute__((arm_streaming, arm_shared_za))
  void svmla_lane_za32[_f32]_vg1x4(uint32_t slice, svfloat32x4_t zn,
                                   svfloat32_t zm, uint64_t imm_idx);
  
  
  // Variants are available for:
  //   _za32[_f32]
  //   _za64[_f64] (only if __ARM_FEATURE_SME_F64F64 != 0)
  __attribute__((arm_streaming, arm_shared_za))
  void svmls_lane_za32[_f32]_vg1x2(uint32_t slice, svfloat32x2_t zn,
                                   svfloat32_t zm, uint64_t imm_idx);
  
  
  // Variants are available for:
  //   _za32[_f32]
  //   _za64[_f64] (only if __ARM_FEATURE_SME_F64F64 != 0)
  __attribute__((arm_streaming, arm_shared_za))
  void svmls_lane_za32[_f32]_vg1x4(uint32_t slice, svfloat32x4_t zn,
                                   svfloat32_t zm, uint64_t imm_idx);
  
  ```

#### FMLAL, BFMLAL, SMLAL, UMLAL (single)

Multi-vector multiply-add long (widening)

``` c
  // Variants are also available for _za32[_f16], _za32[_s16] and _za32[_u16]
  __attribute__((arm_streaming, arm_shared_za))

>   void svmla[_single]_za32[_bf16]_vg2x1(uint32_t slice, svbfloat16_t zn,

Reviewer 2: I'm not sure the `vg2x1` and `vg4x1` functions should have a `[_single]` suffix.  I'd understood `[_single]` to be for functions that take a tuple argument followed by a single-vector argument, to distinguish them from two tuple arguments.  That is, `[_single]` seems like it's for `x2`, `x4`, `vgNx2` and `vgNx4`.

  
  ```

#### SQRSHR, UQRSHR

Multi-vector saturating rounding shift right narrow

``` c
  // Variants are also available for _u16_x2
  __attribute__((arm_streaming))
  svint16_t svqrshr[_s16_x2](svint32x2_t zn, uint64_t imm);
  

  // Variants are also available for _u8_x4
  __attribute__((arm_streaming))
  svint8_t svqrshr[_s8_x4](svint32x4_t zn, uint64_t imm);
  

  // Variants are also available for _u16_x4
  __attribute__((arm_streaming))
  svint16_t svqrshr[_s16_x4](svint64x4_t zn, uint64_t imm);
  
  
  ```

#### SQRSHRN, UQRSHRN

Multi-vector saturating rounding shift right narrow and interleave

``` c
  // Variants are also available for _u16_x2
  __attribute__((arm_streaming_compatible))
  svint16_t svqrshrn[_s16_x2](svint32x2_t zn, uint64_t imm);
  
  
  // Variants are also available for _u8_x4
  __attribute__((arm_streaming))
  svint8_t svqrshrn[_s8_x4](svint32x4_t zn, uint64_t imm);

  // Variants are also available for _u16_x4
  __attribute__((arm_streaming))
  svint16_t svqrshrn[_s16_x4](svint64x4_t zn, uint64_t imm);
  
  ```

#### SQRSHRU

Multi-vector saturating rounding shift right unsigned narrow

``` c
  
  __attribute__((arm_streaming))

>   svuint16_t svsqrshru[_u16_x2](svint32x2_t zn, uint64_t imm);

Reviewer 2: Not saying this is obviously the best choice, but the convention in `arm_neon.h` is to:
* drop the `s`
* use a signed suffix rather than unsigned suffix, reflecting the source type rather than the destination type

SVE2 followed the same approach, for consistency.  So I think this should be `svqrshru[_s32_x2]`.

  
  
  __attribute__((arm_streaming))
  svuint8_t svsqrshru[_u8_x4](svint32x4_t zn, uint64_t imm);

  __attribute__((arm_streaming))
  svuint16_t svsqrshru[_u16_x4](svint64x4_t zn, uint64_t imm);
  
  ```

#### SQRSHRUN

Multi-vector saturating rounding shift right unsigned narrow and interleave

``` c
  
  __attribute__((arm_streaming_compatible))
  svuint16_t svsqrshrun[_u16_x2](svint32x2_t zn, uint64_t imm);
  
  
  // Variants are also available for _u16
  __attribute__((arm_streaming))
  svuint8_t svsqrshrun[_u8_x4](svint32x4_t zn, uint64_t imm);
  
  ```

#### SQDMULH (single)

Multi-vector signed saturating doubling multiply high

``` c
  // Variants are also available for _single_s16_x2, _single_s32_x2
  // and _single_s64_x2
  __attribute__((arm_streaming))

>   svint8x2_t svsqdmulh[_single_s8_x2](svint8x2_t zdn, svint8_t zm);

Reviewer 2: I think we should drop the `s` prefix, for consistency with Neon/SVE2, and with the S/U operations above.  Same for SQDMULH.

  
  
  // Variants are also available for _single_s16_x4, _single_s32_x4
  // and _single_s64_x4
  __attribute__((arm_streaming))
  svint8x4_t svsqdmulh[_single_s8_x4](svint8x4_t zdn, svint8_t zm);
  
  ```

#### SQDMULH (multi)

Multi-vector signed saturating doubling multiply high

``` c
  // Variants are also available for _s16_x2, _s32_x2 and _s64_x2
  __attribute__((arm_streaming))
  svint8x2_t svsqdmulh[_s8_x2](svint8x2_t zdn, svint8x2_t zm);
  
  
  // Variants are also available for _s16_x4, _s32_x4 and _s64_x4
  __attribute__((arm_streaming))
  svint8x4_t svsqdmulh[_s8_x4](svint8x4_t zdn, svint8x4_t zm);
  
  ```

#### WHILEGE, WHILEGT, WHILEHI, WHILEHS, WHILELE, WHILELO, WHILELS, WHILELT

While (resulting in predicate-as-counter). ``vl`` is expected to be 2 or 4.

``` c
  // Variants are also available for _c16, _c32 and _c64
  __attribute__((arm_streaming))
  svcount_t svwhilege_c8(int64_t rn, int64_t rm, uint64_t vl);
  
  
  // Variants are also available for _c16, _c32 and _c64
  __attribute__((arm_streaming))
  svcount_t svwhilegt_c8(int64_t rn, int64_t rm, uint64_t vl);
  
  
  // Variants are also available for _c16, _c32 and _c64
  __attribute__((arm_streaming))

>   svcount_t svwhilehi_c8(uint64_t rn, uint64_t rm, uint64_t vl);

Reviewer 2: For SVE, we instead used le/lt/ge/gt for unsigned inputs too, with a suffix to indicate the signedness and size of the values being compared.  This follows the `arm_neon.h` convention of using `vcle` for `<=` between floats, signed integers, and unsigned integers.

  __attribute__((arm_streaming_compatible))
  svboolx2_t svwhilele_b8_x2(int64_t rn, int64_t rm);
  
  
  // Variants are also available for _b16, _b32 and _b64
  __attribute__((arm_streaming_compatible))
  svboolx2_t svwhilelo_b8_x2(uint64_t rn, uint64_t rm);
  
  
  // Variants are also available for _b16, _b32 and _b64
  __attribute__((arm_streaming_compatible))
  svboolx2_t svwhilels_b8_x2(uint64_t rn, uint64_t rm);
  
  
  // Variants are also available for _b16, _b32 and _b64
  __attribute__((arm_streaming_compatible))
  svboolx2_t svwhilelt_b8_x2(int64_t rn, int64_t rm);
  
  ```

#### SUNPK, UUNPK

Multi-vector pack/unpack

``` c
  // Variants are also available for _u16_x2, _u32_x2, _s32_x2, _u64_x2 and _s64_x2
  __attribute__((arm_streaming))
  svint16x2_t svunpk[_s16_x2](svint8_t zn);
  
  
  // Variants are also available for _u16_x4, _u32_x4, _s32_x4, _u64_x4 and _s64_x4
  __attribute__((arm_streaming))
  svint16x4_t svunpk[_s16_x4](svint8x2_t zn);
  
  ```

#### ZIP

Multi-vector zip.

``` c
  // Variants are also available for _u8, _u16, _s16, _f16, _bf16, _u32, _s32, _f32,
  // _u64, _s64 and _f64
  __attribute__((arm_streaming))

>   svint8x2_t svzip[_s8]_x2(svint8_t zn, svint8_t zm);

Reviewer 2: This might be rehashing an old conversation, sorry, but why do these functions take individual vectors rather than x2 and x4 modes?  If the program does have individual vectors, it would be relatively simple to wrap them in `svcreate2` or `svcreate4` before passing them to the function.  But if the program already has a x2 or x4 mode, it would be a bit cumbersome to take them apart for the sake of the call.
